File: go/tests/integration_contracts_test.go
// integration_contracts_test.go
//
// Verifies CircleAI.Integration/Contracts.cs port (integration_contracts.go):
// the record structs carry the right fields, the interface guards hold via the
// in-memory/HTTP-backed impls, and the JSON scalar helpers behave like the C#
// Get*/TryGetProperty they model.

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// Compile-time interface satisfaction across every ported impl.
func TestIntegration_InterfaceGuards(t *testing.T) {
	tr := circleai.NewFakeCarrierTransport()
	var _ circleai.ICalendarConnector = mustCalDav(t, tr)
	var _ circleai.IEmailConnector = mustGmail(t, tr)
	var _ circleai.INewsSource = mustRss(t, tr)
	var wp circleai.IWeatherProvider
	wp, _ = circleai.NewOpenMeteoWeatherProvider(tr)
	if wp.ProviderID() != "open-meteo" {
		t.Fatalf("weather provider id = %q", wp.ProviderID())
	}
	var rp circleai.IRoutingProvider
	rp, _ = circleai.NewOsrmRoutingProvider(tr, circleai.OsrmOptions{})
	if rp.ProviderID() != "osrm" {
		t.Fatalf("routing provider id = %q", rp.ProviderID())
	}
	var ha circleai.IHomeAutomationConnector
	ha, _ = circleai.NewHomeAssistantConnector(tr, circleai.HomeAssistantOptions{BaseURL: "http://ha/", AccessToken: "t"})
	if ha.ProviderID() != "home-assistant" {
		t.Fatalf("home id = %q", ha.ProviderID())
	}
}

func TestIntegration_CalendarEventFields(t *testing.T) {
	desc := "d"
	loc := "l"
	start := time.Date(2026, 7, 11, 9, 0, 0, 0, time.UTC)
	ev := circleai.CalendarEvent{
		EventID: "e1", CalendarID: "c1", Title: "t", Description: &desc, Location: &loc,
		StartUtc: start, EndUtc: start.Add(time.Hour), IsAllDay: false, Attendees: []string{"a@x"},
	}
	if ev.EventID != "e1" || *ev.Description != "d" || *ev.Location != "l" ||
		!ev.StartUtc.Equal(start) || len(ev.Attendees) != 1 {
		t.Fatalf("calendar event fields wrong: %+v", ev)
	}
}

func TestIntegration_EmailAndNewsAndWeatherFields(t *testing.T) {
	em := circleai.EmailMessage{MessageID: "m", From: "f", To: []string{"t"}, Subject: "s", BodyText: "b", Unread: true, Labels: []string{"INBOX"}}
	if em.MessageID != "m" || !em.Unread || em.To[0] != "t" {
		t.Fatalf("email fields wrong: %+v", em)
	}
	ni := circleai.NewsItem{ItemID: "i", SourceID: "src", Title: "ti", Summary: "su", URL: "https://x/y", Tags: []string{"tag"}}
	if ni.URL != "https://x/y" || ni.Tags[0] != "tag" {
		t.Fatalf("news fields wrong: %+v", ni)
	}
	ws := circleai.WeatherSample{TempC: 21.5, FeelsLikeC: 20, PrecipMm: 0, WindKph: 10, CloudPct: 30, Condition: "clear sky"}
	if ws.TempC != 21.5 || ws.CloudPct != 30 || ws.Condition != "clear sky" {
		t.Fatalf("weather fields wrong: %+v", ws)
	}
	re := circleai.RouteEstimate{DistanceKm: 5, Duration: 10 * time.Minute, Polyline: []circleai.GeoPoint{{Lat: 1, Lon: 2}}}
	if re.DistanceKm != 5 || re.Duration != 10*time.Minute || re.Polyline[0].Lat != 1 {
		t.Fatalf("route fields wrong: %+v", re)
	}
	he := circleai.HaEntity{EntityID: "light.k", FriendlyName: "Kitchen", Domain: "light", State: "on", Attributes: map[string]string{"brightness": "128"}}
	if he.Domain != "light" || he.Attributes["brightness"] != "128" {
		t.Fatalf("ha entity fields wrong: %+v", he)
	}
}

// --- shared test helpers / token providers ---

func fixedToken(tok string) circleai.AccessTokenProvider {
	return func(ctx context.Context) (string, error) { return tok, nil }
}

func mustCalDav(t *testing.T, tr *circleai.FakeCarrierTransport) *circleai.CalDavCalendarConnector {
	t.Helper()
	c, err := circleai.NewCalDavCalendarConnector(tr, circleai.CalDavCalendarOptions{
		CalendarURI: "https://cal.example.com/dav/user/calendars/personal/", Username: "u", Password: "p",
	})
	if err != nil {
		t.Fatalf("new caldav: %v", err)
	}
	return c
}

func mustGmail(t *testing.T, tr *circleai.FakeCarrierTransport) *circleai.GmailEmailConnector {
	t.Helper()
	c, err := circleai.NewGmailEmailConnector(tr, circleai.GmailOptions{AccessTokenProvider: fixedToken("tok")})
	if err != nil {
		t.Fatalf("new gmail: %v", err)
	}
	return c
}

func mustRss(t *testing.T, tr *circleai.FakeCarrierTransport) *circleai.RssNewsSource {
	t.Helper()
	c, err := circleai.NewRssNewsSource(tr, circleai.RssOptions{FeedURL: "https://news.example.com/rss.xml"})
	if err != nil {
		t.Fatalf("new rss: %v", err)
	}
	return c
}

File: typescript/tests/integration_contracts.test.ts
// integration_contracts.test.ts
// Verifies the CircleAI.Integration (Contracts.cs) port: record factories, the
// DateTimeOffset.MinValue sentinel, the IHttpClient transport helpers
// (ensureSuccess / isSuccessStatusCode / resolveUrl), and the string helpers.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  DateTimeOffsetMinValue,
  isMinValue,
  ensureSuccess,
  isSuccessStatusCode,
  resolveUrl,
  isNullOrWhiteSpace,
  isNullOrEmpty,
  HttpResponseError,
  calendarEvent,
  emailMessage,
  newsItem,
  weatherSample,
  routeEstimate,
  haEntity,
} from "../src/integration/index";

describe("CircleAI.Integration — DateTimeOffset.MinValue", () => {
  it("matches C# DateTimeOffset.MinValue epoch ms and isMinValue", () => {
    assert.equal(DateTimeOffsetMinValue.getTime(), -62135596800000);
    assert.equal(isMinValue(new Date(-62135596800000)), true);
    assert.equal(isMinValue(new Date("2026-01-01T00:00:00Z")), false);
  });
});

describe("CircleAI.Integration — HTTP transport helpers", () => {
  it("isSuccessStatusCode is true only for 2xx", () => {
    assert.equal(isSuccessStatusCode(200), true);
    assert.equal(isSuccessStatusCode(204), true);
    assert.equal(isSuccessStatusCode(299), true);
    assert.equal(isSuccessStatusCode(300), false);
    assert.equal(isSuccessStatusCode(404), false);
    assert.equal(isSuccessStatusCode(500), false);
  });

  it("ensureSuccess returns 2xx responses and throws HttpResponseError otherwise", () => {
    const ok = { statusCode: 200, body: "hi" };
    assert.equal(ensureSuccess(ok), ok);
    assert.throws(() => ensureSuccess({ statusCode: 404, body: "" }), HttpResponseError);
    try {
      ensureSuccess({ statusCode: 503, body: "" });
      assert.fail("expected throw");
    } catch (e) {
      assert.ok(e instanceof HttpResponseError);
      assert.equal(e.statusCode, 503);
    }
  });

  it("resolveUrl honors absolute URLs, relative combine, and leading-slash replace", () => {
    // absolute wins
    assert.equal(resolveUrl("https://base.example/x/", "https://other.example/y"), "https://other.example/y");
    // no base → return as-is
    assert.equal(resolveUrl(undefined, "me/events"), "me/events");
    // relative combine against a directory base
    assert.equal(resolveUrl("https://graph.example/v1.0/", "me/events"), "https://graph.example/v1.0/me/events");
    // base without trailing slash gets one appended
    assert.equal(resolveUrl("https://graph.example/v1.0", "me/events"), "https://graph.example/v1.0/me/events");
    // leading slash replaces the base path
    assert.equal(resolveUrl("https://cal.example/dav/user/cal/", "/foo"), "https://cal.example/foo");
    // .ics combine (CalDAV create)
    assert.equal(
      resolveUrl("https://cal.example/dav/user/cal/", "abc123.ics"),
      "https://cal.example/dav/user/cal/abc123.ics",
    );
  });
});

describe("CircleAI.Integration — string helpers", () => {
  it("isNullOrWhiteSpace / isNullOrEmpty match C# semantics", () => {
    assert.equal(isNullOrWhiteSpace(null), true);
    assert.equal(isNullOrWhiteSpace(undefined), true);
    assert.equal(isNullOrWhiteSpace("   "), true);
    assert.equal(isNullOrWhiteSpace(""), true);
    assert.equal(isNullOrWhiteSpace(" x "), false);
    assert.equal(isNullOrEmpty(""), true);
    assert.equal(isNullOrEmpty(null), true);
    assert.equal(isNullOrEmpty("   "), false); // whitespace is NOT empty
    assert.equal(isNullOrEmpty("x"), false);
  });
});

describe("CircleAI.Integration — record factories", () => {
  it("build every record positionally", () => {
    const ev = calendarEvent("e1", "c1", "T", "d", "L", new Date(0), new Date(1000), false, ["a@x"]);
    assert.equal(ev.eventId, "e1");
    assert.equal(ev.description, "d");
    assert.deepEqual(ev.attendees, ["a@x"]);

    const em = emailMessage("m1", "f@x", ["t@x"], "S", "B", new Date(0), true, ["INBOX"]);
    assert.equal(em.unread, true);
    assert.deepEqual(em.to, ["t@x"]);

    const ni = newsItem("i1", "s1", "T", "sum", "https://x/", new Date(0), ["tag"]);
    assert.equal(ni.url, "https://x/");
    assert.deepEqual(ni.tags, ["tag"]);

    const ws = weatherSample(new Date(0), 20, 19, 0.5, 12, 40, "rain");
    assert.equal(ws.condition, "rain");
    assert.equal(ws.windKph, 12);

    const re = routeEstimate(5.5, 600000, [{ lat: 1, lon: 2 }]);
    assert.equal(re.distanceKm, 5.5);
    assert.deepEqual(re.polyline[0], { lat: 1, lon: 2 });

    const ha = haEntity("light.k", "Kitchen", "light", "on", new Map([["brightness", "255"]]));
    assert.equal(ha.domain, "light");
    assert.equal(ha.attributes.get("brightness"), "255");
  });
});
